### PR TITLE
Adding Lyrics Support

### PR DESCRIPTION
**Pull request by rhnvrm**
screenshot from 2015-07-18 12 32 52

Hi,

I've added a module for fetching lyrics from lyrics.wikia.com which is currently logging to the console. I've also added a lyrics box in place of the new playlist button which I have moved infront of the 'playlist' label. Currently it is above the songname and currently has placeholder text in the image. I'm currently running into trouble with socket.io and am unable to send back the lyrics so as to update them there. 

Obviously this needs work but hopefully, if I could be guided on how to integrate this it would prove to be a cool feature.

### CODE REVIEW

**Comment by benkaiser** [comment]
Hi @rhnvrm,

First of all, awesome work! I think that's an awesome idea adding lyrics support, and also how you moved the add playlists button, it looks way nicer there!

I'll have to review this later this week or maybe next week as I'm flat out and trying to enjoy myself on a vacation at the moment, but I look forward to reviewing your code and suggesting some improvements.

Cheers,
Ben

**Comment by rhnvrm** [comment]
Sure! No problem.

**Comment by benkaiser** [comment]
Okay, I have time to review this now, apologies for the wait.

First of all, when running, neither `cheerio` or `underscore` have been added to the `package.json`, you can add them automatically when installing them by passing the `--save` flag.

Secondly, the app crashes on use:
```
/home/benkaiser/GIT/node-music-player/node_modules/request/index.js:46
    throw new Error('undefined is not a valid uri or options object.')
          ^
Error: undefined is not a valid uri or options object.
    at request (/home/benkaiser/GIT/node-music-player/node_modules/request/index.js:46:11)
    at getLyrics (/home/benkaiser/GIT/node-music-player/scraper.js:54:3)
    at Request._callback (/home/benkaiser/GIT/node-music-player/scraper.js:111:9)
    at Request.self.callback (/home/benkaiser/GIT/node-music-player/node_modules/request/request.js:198:22)
    at Request.emit (events.js:110:17)
    at Request.<anonymous> (/home/benkaiser/GIT/node-music-player/node_modules/request/request.js:1057:14)
    at Request.emit (events.js:129:20)
    at IncomingMessage.<anonymous> (/home/benkaiser/GIT/node-music-player/node_modules/request/request.js:1003:12)
    at IncomingMessage.emit (events.js:129:20)
    at _stream_readable.js:908:16
```

I'll let you work out the root cause of that.

One other thing I noticed is that you are hitting the API and retrieving the HTML data and parsing it with `cheerie`... Wouldn't it be simpler to call the wikia api with `&fmt=json` to get JSON data that can simply be parsed using `JSON.parse(data)` in javascript?
http://api.wikia.com/wiki/LyricWiki_API

I'll do a line-by-lin review now and comment any other notes for improvement that I find.

File: routes/index.js [hunk 1]
```diff
@@ -9,6 +9,8 @@ var request = require('request').defaults({ encoding: null });
 var fs = require('fs');
 var path = require('path');
 var MobileDetect = require('mobile-detect');
+var lyrics = require(__dirname + '/../scraper.js');
```
Comment (benkaiser) [inline routes/index.js]: Perhaps a different file to `scraper.js`? Maybe `lyric_fetcher.js` or simply `lyrics.js`.

File: scraper.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  return lyrics;
+};
+    
+// get the lyrics for an individual song 
+var getLyrics = function(url, callback) {
+  request(url, function(error, response, html) {
+    if (!error) {
+      var $ = cheerio.load(html);
+      $('script').remove()
+      var cleanlyrics = cleanLyrics($('.lyricbox').html());
```
Comment (benkaiser) [inline scraper.js]: Back to my overarching comment, json seems like it would be a lot easier to process.

File: scraper.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      callback(cleanlyrics);
+    }
+  });
+};
+
+
+exports.findLyrics2 = function(artist){
```
Comment (benkaiser) [inline scraper.js]: Where is this function used? What is it's purpose?

File: scraper.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  request(url, function(error, response, html) {
+    if (!error) {
+      var $ = cheerio.load(html);
+      var songs = $('.songs').find('li');
+      for (var i = 0; i < songs.length; i++) {
+        
+        song_name = ((($(songs[i]).find('a').attr('href')).split(':')[2]).toLowerCase().split('_'));
+
+        //song_name = song.split('_');
+        //console.log(song_name);
```
Comment (benkaiser) [inline scraper.js]: committing commented out `console.log` statements seems like a bad practice... just remove them?

File: scraper.js (hunk 2 above)
Comment (rhnvrm) [inline scraper.js]: Actually, we are scraping the URLs in the li tag at this page http://lyrics.wikia.com/api.php?artist=owl%20city and getting the lyrics by calling request and only using the .lyricbox element's content from those URLs. 
Using JSON here would make it difficult as . 
http://lyrics.wikia.com/api.php?artist=owl%20city&fmt=json

File: routes/index.js (hunk 1 above)
Comment (rhnvrm) [inline routes/index.js]: changed to lyrics.js

File: scraper.js (hunk 3 above)
Comment (rhnvrm) [inline scraper.js]: Explained in my previous comment. It is used inside the exported findLyrics2 function.

File: scraper.js (hunk 4 above)
Comment (rhnvrm) [inline scraper.js]: Sorry for that. Fixed.

**Comment by rhnvrm** [comment]
Hi, I've fixed the things you mentioned and have explained why I am not using JSON in the comments above. 

I am unable to recreate the error you are getting... 

Also, please suggest how I should go on to show the lyrics in the frontend?

**Comment by benkaiser** [comment]
I actually really disagree with wikia here. I looked at their documentation and in the [overview section](http://api.wikia.com/wiki/LyricWiki_API) it states:

> This process is different from the original API which used to return the full lyrics in the API result. This had to be changed due to licensing restrictions.

Due to this I am not interested in the music player displaying partial lyrics.

And the solution you have provided, although rather awesome in how it works to show the full lyrics, they are not legally allowed to release the full lyrics via API and using a scraper (which is essentially what your solution does) is a very messy solution.

On a side note, they do have a `realjson` format that outputs proper json that is easy to parse like so:
```
request('http://lyrics.wikia.com/api.php?artist=owl%20city&title=fireflies&fmt=realjson', function(error, response, body) {
  var cleanLyrics = JSON.parse(body);
  console.log(cleanLyrics.lyrics);
});
```

Perhaps one cool task if you are interested in scraping, would be to scrape every single lyric from their website and create your own open API for the tracks. I'll donate you some cpu cycles on one of my servers if you want to host that solution somewhere :stuck_out_tongue_winking_eye:

**Comment by benkaiser** [comment]
Or... if you package all this messy functionality into a third-party module that takes title and artist and then does all this messy logic in the background, then just include that in my application. I would be happy with merging that too. (this is how I dealt with my soundcloud-resolver module, which was created by myself but is a separate module to the music player).

**Comment by rhnvrm** [comment]
I've made the API and hosted it on Heroku. https://github.com/rhnvrm/lyric-api

File: static/js/player/player.js [hunk 5]
```diff
@@ -215,6 +215,13 @@ function PlayState(){
     // show the songs info
     info = new InfoView();
     MusicApp.infoRegion.show(info);
+    // update the lyrics panel
+    song_details = {artist:this.current_song.attributes.display_artist, title:this.current_song.attributes.title};
+    socket.emit('fetch_lyrics', song_details);
+    //lyrics = this.current_song.attributes.title;
+    socket.on('lyrics_result', function(data) {
+      $('.sidebar_lyrics').html('data');
```
Comment (rhnvrm) [inline static/js/player/player.js]: Can you help with this area of code? I really am unable to find a way to update the lyrics...

**Comment by rhnvrm** [comment]
Also, should we save the lyrics in the ID3 tags?

**Comment by benkaiser** [comment]
Hey @rhnvrm , that api is awesome! thanks for throwing that together!

 I've made some changes and pushed them to the [rhnvrm-master branch](https://github.com/benkaiser/node-music-player/tree/rhnvrm-master). There are only a few more changes / things I'd like you to think about before I merge.

1. Return the string from the API without quotes, they currently all have double quotes around them. A nicer way would be to send JSON from your API, and then to parse that JSON on the Node Music Player end. Would be good if you also had some way of communicating back to node music player if you were unable to find the lyrics (then if we need to, we could handle this differently).
2. Have a think about how the lyrics box is very small. From a user perspective it would be nice if they could resize it. Is this possible with something like css flexbox maybe? You don't have to implement this, but I'd like you to suggest some options and I can work on them in the future :)
3. Also have a think about, and write in a response here what your plan is regarding storing these for the songs? it would be awesome to write them at least to the Node Music Player database, and then check there before sending a request off again later.

On your server also, do you have any caching in place? So that if the same item is requested 10000 times it only hits the wikia servers once? That would be useful to ensure that if many many people use this feature in node music player, your server can handle it :)

**Comment by rhnvrm** [comment]
For the lyricbox, I am not sure on how we should proceed.

Maybe we could use http://dyndiv.markusbordihn.de/example/moveable-resizeable/

I also found this snippet which might be of use if we add jQuery to the project:
$("#element_to_resize").resizable({ handles: 'n' }).bind("resize", function (e, ui) {
            $(this).css("top", "auto");
        });

**Comment by rhnvrm** [comment]
I don't have caching set up but I'm reading on how to implement it. Saving to the Node Player DB would be good. We could also add it to the file itself.

**Comment by ohnx** [comment]
@rhnvrm Personally, adding the lyrics into the database is better, as adding the lyrics to the file tags makes the file itself bigger.

**Comment by xtrm0** [comment]
@ohnx It's a < 1kb increase on a file that probably has > 3mb.

**Comment by ohnx** [comment]
@xtrm0 nonetheless, the database is probably more appropriate for lyrics as it's mainly this app that uses it. Also, I not keen on the app modifying my music files

**Comment by benkaiser** [comment]
Chill guys, I'm going to have to side with xtrm0 here since there is an id3 tag for lyrics. The issue here as that the id3 tag libraries in use are rather limited in functionality when it comes to some things and I'm not sure if we can technically both read and write the lyrics tag. But if we can I would like to do that.

Would someone like to fork the branch I created and add the saving to db and file operations?

Also @rhnvrm how did you go at adding caching support?
I'll try and look into the resizing stuff this weekend and see how I go.

**Comment by rhnvrm** [comment]
Hosting the API on the free heroku platform is too difficult as it just emailed me that the free dyno requires a recharge plus I cant setup any free CDN with the free platform there. Before we integrate it to the main branch of the node player, if anyone could donate and host the API for me and integrate it with my git repo's pushes and apply cdn on it would be amazing.

**Comment by benkaiser** [comment]
@rhnvrm have you ever heard of OpenShift by redhat? They have a free tier that allows for 24/7 hosting but its just a low powered 'gear' (their equivalent if a dyno).
Might be worth looking into.

All else fails I can host it in my server (the one I use to host the demo of node music player), but I mess around with things on there so I don't consider it highly reliable.

**Comment by ohnx** [comment]
Why not have the lyrics API built into node-music-player?

**Comment by benkaiser** [comment]
@ohnx I'm fine with that, but I'd want it as a separate module that is just included in.

I think it's nice as an API though, because:
1. Other OOS that isn't based on Nodejs can use it.
2. Means node music player only has to retrieve a minimal amount of data (1 request and just the lyrics), vs the several requests and other data in the lyrics code.

**Comment by ohnx** [comment]
@benkaiser the lyrics API that @rhnvrm created is really just a wrapper for lyrics.wikia.com. Only 1 request is being sent.

**Comment by benkaiser** [comment]
@ohnx ah yes, I'm mistaken I thought it was making two. But still I don't mind the abstraction. I'm honestly okay with the player making the request, but I'd want it as a separate module.

**Comment by ohnx** [comment]
@benkaiser I can try do do that.

**Comment by ohnx** [comment]
@benkaiser [lyrics-layer](https://github.com/ohnx/lyrics-layer) might be able to do it.

**Comment by benkaiser** [comment]
@ohnx looks good. Want to fork the branch we currently have and integrate it with that library?

**Comment by ohnx** [comment]
@benkaiser I'll see what I can do.

**Comment by rhnvrm** [comment]
I found out that this won't be legal to introduce in this open source project. Sorry for the delayed response but I think we should close this pull request.

**Comment by benkaiser** [comment]
@rhnvrm you do realise that you can download songs directly from youtube and soundcloud in this player? I hardly think adding lyrics support is a concern.

Anyway I might pick this up later and add it in, it would be a really nice feature to have.

**Comment by rhnvrm** [comment]
@benkaiser I'm also not giving up on this project. I'll soon send a new pull request with a better module/api to scrape lyrics.

**Comment by benkaiser** [comment]
Okay sure thing @rhnvrm

**Pull request closed without merging**